File: streampark-common/src/test/scala/org/apache/streampark/common/util/FileUtilsTest.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.streampark.common.util

import org.mockito.MockitoSugar._
import org.scalatest.funsuite.AnyFunSuite

import java.io._

class FileUtilsTest extends AnyFunSuite {

  test("isJarFileType should return true for a valid jar file") {
    val jarFileContent = Array[Byte](0x50, 0x4B, 0x03, 0x04)
    val jarInputStream = new ByteArrayInputStream(jarFileContent)
    assert(FileUtils.isJarFileType(jarInputStream))
  }

  test("isJarFileType should return false for a non-jar file") {
    val textFileContent = Array[Byte](0x54, 0x45, 0x53, 0x54)
    val textInputStream = new ByteArrayInputStream(textFileContent)
    assert(!FileUtils.isJarFileType(textInputStream))
  }

  test("exists should return true if the file exists") {
    val existingFile = new File("existing_file.txt")
    existingFile.createNewFile()
    assert(FileUtils.exists(existingFile))
    existingFile.delete()
  }

  test("exists should return false if the file does not exist") {
    val nonExistingFile = new File("non_existing_file.txt")
    assert(!FileUtils.exists(nonExistingFile))
  }

  test("createTempDir should create a new temporary directory") {
    val tempDir = FileUtils.createTempDir()
    assert(tempDir.exists())
    assert(tempDir.isDirectory)
  }

  test("getPathFromEnv should return the absolute path of the specified environment variable") {
    val envVar = "TEMP"
    val path = FileUtils.getPathFromEnv(envVar)
    assert(path != null && path.nonEmpty)
    val file = new File(path)
    assert(file.exists())
  }

  test(
    "getPathFromEnv should throw IllegalArgumentException if the specified environment variable is not set") {
    val nonExistingEnvVar = "NON_EXISTING_ENV_VAR"
    assertThrows[IllegalArgumentException] {
      FileUtils.getPathFromEnv(nonExistingEnvVar)
    }
  }

  test("resolvePath should throw IllegalArgumentException if the file does not exist") {
    val nonExistingParent = "/tmp"
    val child = "child.txt"

    // Mock file and its behavior
    val mockFile = mock[File]
    when(mockFile.exists).thenReturn(false)

    assertThrows[IllegalArgumentException] {
      FileUtils.resolvePath(nonExistingParent, child)
    }
  }

  test("getSuffix should throw IllegalArgumentException if the filename is null") {
    val filename: String = null
    assertThrows[IllegalArgumentException] {
      FileUtils.getSuffix(filename)
    }
  }

  test("readInputStream should read data from the input stream into the byte array") {
    val inputData = "Hello"
    val inputStream = new ByteArrayInputStream(inputData.getBytes("UTF-8"))
    val byteArray = new Array[Byte](inputData.length)
    FileUtils.readInputStream(inputStream, byteArray)
    val result = new String(byteArray, "UTF-8")
    assert(result == inputData)
  }

  test("readFile should read the content of the file into a string") {
    val content = "Hello"
    val file = new File("test_file.txt")
    val outputStream = new FileOutputStream(file)
    outputStream.write(content.getBytes("UTF-8"))
    outputStream.close()
    val result = FileUtils.readFile(file)
    assert(result == content)
    file.delete()
  }

  test("readEndOfFile should read the last part of a file") {
    val content = "1234567890"
    val file = new File("test_file.txt")
    val outputStream = new FileOutputStream(file)
    outputStream.write(content.getBytes("UTF-8"))
    outputStream.close()
    val result = FileUtils.readEndOfFile(file, 5)
    assert(result.sameElements("67890".getBytes("UTF-8")))
    file.delete()
  }

  test(
    "readEndOfFile should read the entire file if the file size is less than the specified limit") {
    val content = "1234567890"
    val file = new File("test_file.txt")
    val outputStream = new FileOutputStream(file)
    outputStream.write(content.getBytes("UTF-8"))
    outputStream.close()
    val result = FileUtils.readEndOfFile(file, 15)
    assert(result.sameElements(content.getBytes("UTF-8")))
    file.delete()
  }

  test("readFileFromOffset should read the content of the file from a specified offset") {
    val content = "1234567890"
    val file = new File("test_file.txt")
    val outputStream = new FileOutputStream(file)
    outputStream.write(content.getBytes("UTF-8"))
    outputStream.close()
    val result = FileUtils.readFileFromOffset(file, 5, 5)
    assert(result.sameElements("67890".getBytes("UTF-8")))
    file.delete()
  }

  test(
    "readFileFromOffset should throw IllegalArgumentException if the startOffset is greater than the file length") {
    val file = new File("test_file.txt")
    assertThrows[IllegalArgumentException] {
      FileUtils.readFileFromOffset(file, 15, 5)
    }
  }

  test("tailOf should return null if the file does not exist") {
    val result = FileUtils.tailOf("/nonexistent_file.txt", 0, 5)
    assert(result == null)
  }
}

File: streampark-common/src/test/java/org/apache/streampark/common/util/FileUtilsTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.common.util;

import org.junit.jupiter.api.Test;

import java.io.ByteArrayInputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.nio.charset.StandardCharsets;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

class FileUtilsTest {

    @Test
    void isJarFileTypeShouldReturnTrueForValidJarHeader() {
        byte[] jarFileContent = new byte[]{0x50, 0x4B, 0x03, 0x04};
        assertTrue(FileUtils.isJarFileType(new ByteArrayInputStream(jarFileContent)));
    }

    @Test
    void isJarFileTypeShouldReturnFalseForNonJarHeader() {
        byte[] textFileContent = new byte[]{0x54, 0x45, 0x53, 0x54};
        assertFalse(FileUtils.isJarFileType(new ByteArrayInputStream(textFileContent)));
    }

    @Test
    void existsShouldReturnTrueIfFileExists() throws Exception {
        File existingFile = File.createTempFile("existing", ".txt");
        try {
            assertTrue(FileUtils.exists(existingFile));
        } finally {
            existingFile.delete();
        }
    }

    @Test
    void existsShouldReturnFalseIfFileDoesNotExist() {
        assertFalse(FileUtils.exists(new File("non_existing_file.txt")));
    }

    @Test
    void createTempDirShouldCreateDirectory() {
        File tempDir = FileUtils.createTempDir();
        assertTrue(tempDir.exists());
        assertTrue(tempDir.isDirectory());
        tempDir.delete();
    }

    @Test
    void getPathFromEnvShouldReturnExistingPath() {
        String envName = System.getenv("TMPDIR") != null ? "TMPDIR" : "TEMP";
        if (System.getenv(envName) != null || System.getProperty(envName) != null) {
            String path = FileUtils.getPathFromEnv(envName);
            assertNotNull(path);
            assertFalse(path.isEmpty());
        }
    }

    @Test
    void getPathFromEnvShouldThrowIfVariableNotSet() {
        assertThrows(NullPointerException.class, () -> FileUtils.getPathFromEnv("NON_EXISTING_ENV_VAR"));
    }

    @Test
    void resolvePathShouldThrowIfParentDoesNotExist() {
        assertThrows(IllegalArgumentException.class,
            () -> FileUtils.resolvePath("/tmp/nonexistent-parent-114514", "child.txt"));
    }

    @Test
    void getSuffixShouldThrowIfFilenameIsNull() {
        assertThrows(NullPointerException.class, () -> FileUtils.getSuffix(null));
    }

    @Test
    void readInputStreamShouldReadBytes() throws Exception {
        String inputData = "Hello";
        byte[] byteArray = new byte[inputData.length()];
        FileUtils.readInputStream(new ByteArrayInputStream(inputData.getBytes(StandardCharsets.UTF_8)), byteArray);
        assertEquals(inputData, new String(byteArray, StandardCharsets.UTF_8));
    }

    @Test
    void readFileShouldReadContent() throws Exception {
        File file = File.createTempFile("read-file", ".txt");
        try {
            try (FileOutputStream outputStream = new FileOutputStream(file)) {
                outputStream.write("Hello".getBytes(StandardCharsets.UTF_8));
            }
            assertEquals("Hello", FileUtils.readFile(file));
        } finally {
            file.delete();
        }
    }

    @Test
    void readEndOfFileShouldReadTail() throws Exception {
        File file = File.createTempFile("read-end", ".txt");
        try {
            try (FileOutputStream outputStream = new FileOutputStream(file)) {
                outputStream.write("1234567890".getBytes(StandardCharsets.UTF_8));
            }
            assertArrayEquals("67890".getBytes(StandardCharsets.UTF_8), FileUtils.readEndOfFile(file, 5));
        } finally {
            file.delete();
        }
    }

    @Test
    void readEndOfFileShouldReadEntireFileWhenLimitExceedsSize() throws Exception {
        File file = File.createTempFile("read-end-all", ".txt");
        try {
            String content = "1234567890";
            try (FileOutputStream outputStream = new FileOutputStream(file)) {
                outputStream.write(content.getBytes(StandardCharsets.UTF_8));
            }
            assertArrayEquals(content.getBytes(StandardCharsets.UTF_8), FileUtils.readEndOfFile(file, 15));
        } finally {
            file.delete();
        }
    }

    @Test
    void readFileFromOffsetShouldReadPartialContent() throws Exception {
        File file = File.createTempFile("read-offset", ".txt");
        try {
            try (FileOutputStream outputStream = new FileOutputStream(file)) {
                outputStream.write("1234567890".getBytes(StandardCharsets.UTF_8));
            }
            assertArrayEquals("67890".getBytes(StandardCharsets.UTF_8), FileUtils.readFileFromOffset(file, 5, 5));
        } finally {
            file.delete();
        }
    }

    @Test
    void readFileFromOffsetShouldThrowWhenOffsetTooLarge() throws Exception {
        File file = File.createTempFile("read-offset-bad", ".txt");
        try {
            try (FileOutputStream outputStream = new FileOutputStream(file)) {
                outputStream.write("123".getBytes(StandardCharsets.UTF_8));
            }
            assertThrows(IllegalArgumentException.class, () -> FileUtils.readFileFromOffset(file, 15, 5));
        } finally {
            file.delete();
        }
    }

    @Test
    void tailOfShouldReturnNullIfFileDoesNotExist() throws Exception {
        assertNull(FileUtils.tailOf("/nonexistent_file.txt", 0, 5));
    }
}
